### PR TITLE
Upgrade Dagger. 

### PR DESCRIPTION
**Pull request by vRallev**
Initially, Dagger relied on the order of parameters for assisted injection and parameter types had to match between the inject constructor and the factory method. Dagger now introduced a new string identifier `@Assisted("identifier")`. For each parameter they compute a key, which is like a hash of the identifier and the type. The order of parameters is no longer important. Rather, they rely on the keys. There should be no duplicate key and for each key in the inject constructor must be a key in the factory function.

For this PR I upgraded Dagger, updated all tests with the behavior changes, added tests for new edge cases and then fixed all implementations to make the tests pass.

### CODE REVIEW

**Comment by vRallev** [comment]
This whole things felt like interview task with keeping track of elements in multiple lists, diffing and comparing...

**Review by christiandeange — APPROVED** [review]
This is very much a `:rubber-stamp:` – if the tests pass I assume everything is working as expected 😅

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/KotlinPoet.kt [hunk 1]
```diff
@@ -244,14 +258,26 @@ internal fun List<KtCallableDeclaration>.mapToParameter(module: ModuleDescriptor
...
+    val assistedAnnotation = parameter.findAnnotation(assistedFqName)
+    val assistedIdentifier =
+      (assistedAnnotation?.valueArguments?.firstOrNull() as? KtValueArgument)
+        ?.children
+        ?.filterIsInstance<KtStringTemplateExpression>()
+        ?.single()
+        ?.children
+        ?.first()
+        ?.text
+        ?: ""
```
[the remark this answers is not in the document]
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/KotlinPoet.kt:270]: Welcome to PSI!

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedInjectGenerator.kt [hunk 2]
```diff
@@ -185,4 +189,37 @@ internal class AssistedInjectGenerator : PrivateCodeGenerator() {
...
   }
+
+  private fun checkAssistedParametersAreDistinct(
+    clazz: KtClassOrObject,
+    parameters: List<Parameter>
+  ) {
+    // Parameters are identical, if there types and identifier match.
+    val duplicateAssistedParameters = parameters
+      .groupBy { it.assistedParameterKey }
+      .filter { (_, parameterList) ->
```
[the remark this answers is not in the document]
Comment (vRallev) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedInjectGenerator.kt:200]: Nice
